### PR TITLE
WebSocket support close #99

### LINKED ISSUE #99: [feature] WebSocket support
## Background
* Easegress is popular for reverse proxy usage in production and development scenarios.
* In reversed proxy use cases, `WebSocket` is a widely used protocol for a full-duplex communication solution between client and server. It relies on TCP.[1] 
* Many reverse proxy support `WebSocket`,e.g., NGINX[2], Traefik, and so on.

## Proposal 
* Bring `WebSocketServer` as an BusinessController to Easegress. (Should treat it as an `HTTPServer` like traffic gate in the future. )
* Using `github.com/gorilla/websocket` as the `WebSocket` client implementation, since it has rich features supproted and a quite active community.（15k star/2.5k fork) Comparing to the original `golang.org/x/net/websocket` package, it can `receive fragmented message` and `send close message`.[3]

## Design
1. Spec 
```yaml
kind: WebSocketServer
name: websocketSvr
https: false                  # client need to use http/https firstly for connection upgrade      
certBase64:
keyBase64:
port: 10020                   # proxy servers listening port

backend: ws://localhost:3001  # the reserved proxy target
                              #  Easegress will exame the backend URL's scheme, If it starts with `wss`,
                              #  then `wssCerBase64` and `wssKeyBase64` must not be empty

wssCertBase64:                # wss backend certificate in base64 format
wssKeyBase64:                 # wss backend key in base64 format
```
2. Request sequence
```
   +--------------+                +--------------+                +--------------+  
   |              |  1             |              |   2            |              | 
   |   client     +--------------->|  Easegress   +--------------->|  websocket   |
   |              |<---------------+(WebSocketSvr)|<---------------+  backend     | 
   |              |  4             |              |   3            |              |
   +--------------+                +--------------+                +--------------+
```
3. The `WebSocketServer` shares some configurations with `HTTPServer`. Since `WebSocket` runs on `HTTP 1.1` only, and it works on opening two long-life goroutines for the reverse proxy scenario, it's quite different from the original Easegress' HTTPServer+Pipeline's process. So I decide not to implement it into `HTTPServer` and catalog it into `TrafficGate` in the first version right now. 

4. I will implement this `WebSocketServer` first version and try to figure out how to integrate with the original `HTTPServer` at the same time. (Moreover, multiple protocols supporting requirements, such as `MQTT`, `TPC`, are on the way... )

## References
1. https://datatracker.ietf.org/doc/html/rfc6455
2. https://www.nginx.com/blog/websocket-nginx/
3. https://github.com/gorilla/websocket

### PR DESCRIPTION
**Pull request by benja-wu**
1. Introducing WebSocketServer into Easegress

## TODO
1. Considering to merge it into `HTTPServer`, and managed by `TrafficController`.
2. Updating HTTPPipeline for handling this none-one-request-one-response scenario.(Two way long-life connection)

### CODE REVIEW

File: pkg/object/websocketserver/proxy.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+	}
+
+	// defaultDialer is a dialer with all fields set to the default zero values.
+	defaultDialer = websocket.DefaultDialer
+
+	// defaultIntervalSecond is the default interval for polling websocket client and server.
+	defaultIntervalSecond time.Duration = 1
+)
+
+// Proxy is an handler that takes an incoming WebSocket
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/proxy.go:51]: a handler

File: pkg/object/websocketserver/spec.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+import (
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"strings"
+)
+
+type (
+	// Spec describes the HTTPServer.
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/spec.go:28]: WebSocketServer

File: pkg/object/websocketserver/spec.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	Spec struct {
+		Port    uint16 `yaml:"port" jsonschema:"required,minimum=1"`
+		HTTPS   bool   `yaml:"https" jsonschema:"required"`
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/spec.go:31]: HTTPS? Should it be `WSS` or `EnableTLS`?

File: pkg/object/websocketserver/proxy.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+				return
+			}
+		}
+	}
+}
+
+// run runs the websocket proxy.
+func (p *Proxy) run() {
+	spec := p.superSpec.ObjectSpec().(*Spec)
+	backendURL, err := url.Parse(spec.Backend)
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/proxy.go:129]: The `Backend` should be validated in `Spec.Validate` or use `jsonschema:"required,url"

File: pkg/object/websocketserver/proxy.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+	if spec.HTTPS {
+		tlsConfig, err := spec.tlsConfig()
+		if err != nil {
+			logger.Errorf("gen websocketserver's httpserver tlsConfig: %#v, failed: %v", spec, err)
+		}
+		svr.TLSConfig = tlsConfig
+	}
+
+	if err := svr.ListenAndServe(); err != nil {
+		logger.Errorf("websocketserver ListenAndServe failed! err: %s\n", err.Error())
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/proxy.go:164]: `logger.Errorf("websocketserver ListenAndServe failed: %v", err)`

File: pkg/object/websocketserver/proxy.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	defer connClient.Close()
+
+	errClient := make(chan error, 1)
+	errBackend := make(chan error, 1)
+	stop := make(chan struct{})
+
+	defer close(stop)
+
+	// pass msg from backend to client vias WebSocket protocol.
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/proxy.go:248]: via

File: pkg/object/websocketserver/proxy.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+		upgradeHeader.Set("Set-Cookie", hdr)
+	}
+	return upgradeHeader
+}
+
+// handle implements the http.Handler that proxies WebSocket connections.
+func (p *Proxy) handle(rw http.ResponseWriter, req *http.Request) {
+	connBackend, resp, err := p.dialer.Dial(p.buildReq(req).String(), p.copyHeader(req))
+	if err != nil {
+		logger.Errorf("Proxy: couldn't dial to ws remote backend url: %s, err: %v", p.backendURL.String(), err)
```
Comment (xxx7xxxx) [inline pkg/object/websocketserver/proxy.go:217]: `logger.Errorf("%s dial %s failed: %v", p.superSpec().Name(), p.backendURL.String(), err)`. Please keep error message accurate and clean without redundant words.

**Review by xxx7xxxx — COMMENTED** [review]
This is kind of a simple websocket proxy isolated with other components in EG, there left work to naturally merged it into existed traffic ecosystem.

**Comment by benja-wu** [comment]
> This is kind of a simple websocket proxy isolated with other components in EG, there left work to naturally merged it into existed traffic ecosystem.

@xxx7xxxx  Yes, cause I haven't figure out how to update the `TrafficController` and `RawTrafficController` for handling none-HTTP related traffic, such as TCP, WebSocket(After protocol switching, it uses TCP connection directly). And the Pipeline also needs to be updated for handling the none-one-request-one-response scenario. 

We need a big issue and design for this, not only for WebSocket/TCP, but also for MQTT protocol

File: pkg/object/websocketserver/spec.go (hunk 3 above)
Comment (benja-wu) [reply pkg/object/websocketserver/spec.go:31]: Actually, this service can enable TLS in two places, one is the HTTP server it starts with, the other is the backend WebSocket's. 
* This `HTTPS` is for the original HTTP server this proxy works on. (Before protocol switching)
* The backend WebSocket service's TLS configure can be enabled when it starts with `wss` in URL.

File: pkg/object/websocketserver/proxy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (benja-wu) [reply pkg/object/websocketserver/proxy.go]: Updated

File: pkg/object/websocketserver/proxy.go (hunk 7 above)
Comment (benja-wu) [reply pkg/object/websocketserver/proxy.go]: Already update to correct log format. Thx

File: pkg/object/websocketserver/proxy.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+)
+
+// Proxy is a handler that takes an incoming WebSocket
+// connection and proxies it to the backend server.
+type Proxy struct {
+	// server is the HTTPServer
+	server    http.Server
+	superSpec *supervisor.Spec
+
+	// BackendURL URL is the
```
Comment (localvar) [inline pkg/object/websocketserver/proxy.go:59]: the comment is incompelete.

File: pkg/object/websocketserver/proxy.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+func newProxy(superSpec *supervisor.Spec) *Proxy {
+	proxy := &Proxy{
+		superSpec: superSpec,
+		done:      make(chan struct{}),
+	}
+	go proxy.run()
+	return proxy
+}
+
+// buildReq builds an URL with backend in spec and original HTTP request.
```
Comment (localvar) [inline pkg/object/websocketserver/proxy.go:82]:
```suggestion
// buildRequestURL builds an URL with backend in spec and original HTTP request.
```

File: pkg/object/websocketserver/spec.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		Backend string `yaml:"backend" jsonschema:"required"`
+
+		CertBase64 string `yaml:"certBase64" jsonschema:"omitempty,format=base64"`
+		KeyBase64  string `yaml:"keyBase64" jsonschema:"omitempty,format=base64"`
+
+		wssCertBase64 string `yaml:"certBase64" jsonschema:"omitempty,format=base64"`
+		wssKeyBase64  string `yaml:"keyBase64" jsonschema:"omitempty,format=base64"`
```
Comment (localvar) [inline pkg/object/websocketserver/spec.go:39]: is it need to bind multiple certs? as #31

File: pkg/object/websocketserver/spec.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+		cert, err := tls.X509KeyPair(certPem, keyPem)
+		if err != nil {
+			return nil, fmt.Errorf("generate x509 key pair failed: %v", err)
+		}
+		certificates = append(certificates, cert)
+	} else {
+		return nil, fmt.Errorf("cert/key base64 empty, certBase64: %s, keyBase64: %s", spec.wssKeyBase64, spec.wssCertBase64)
+	}
+	return &tls.Config{Certificates: certificates}, nil
+}
```
Comment (localvar) [inline pkg/object/websocketserver/spec.go:97]: propose refactor to eliminate code duplication.

File: pkg/object/websocketserver/spec.go (hunk 10 above)
Comment (benja-wu) [reply pkg/object/websocketserver/spec.go:39]: As the implementation right now:
1. Before protocol switching, there is only one HTTPServer for handling the traffic. 
2. After protocol switching, the `WebScoket` server only accepts one routable backend. 

So I think maybe it's not necessary to introduce multiple certs support in this version.

File: pkg/object/websocketserver/spec.go (hunk 11 above)
Comment (benja-wu) [reply pkg/object/websocketserver/spec.go:97]: Sure, let me update it

**Comment by benja-wu** [comment]
Updated as comments.

**Review by localvar — APPROVED** [review]

**Review by xxx7xxxx — APPROVED** [review]

**Pull request merged**